streams/src/main/java/org/apache/kafka/streams/query/QueryResult.java

```diff
@@ -197,6 +197,18 @@ public R getResult() {
         return result;
     }
 
+    @SuppressWarnings("unchecked")
+    public <V> QueryResult<V> swapResult(final V value) {
+        if (isFailure()) {
+            return (QueryResult<V>) this;
+        } else {
+            final QueryResult<V> result = new QueryResult<>(value);
```

Comment:
The generic type of `QueryResult` is `R` -- it does not seem safe to allow swapping in a `V` result?

If we need to allow for different return types, why does `QueryType` has a generic parameter to begin with? Seems useless if we cannot rely on it?

Reply:
What's happening here is that we're turning a `QueryResult<R>` into a `QueryResult<V>`. A concrete example (in fact the only use case) of this is in the MeteredStore, we get back a raw result from the BytesStore and need to deserialize it, so we need to convert the `QueryResult<byte[]>` into a `QueryResult<Integer>` or something.

Reply:
@vvcephei WDYT about having an extended `QueryResultInBytes` (just a placeholder for name) on `QueryResult<byte[]>` and move this function to that extended class? This way we can avoid mistakenly using the swap functions.

Reply:
Why not use two objects?

Reply:
Thanks, @guozhangwang , I think something like that will be the outcome of this follow-on work: https://issues.apache.org/jira/browse/KAFKA-13526

We'll tackle that question before the first release of this new API.

Reply:
Thanks, @mjsax , I'm not sure precisely what you mean. This does create a new object. If you think it would be clearer to add a constructor allowing people to set the result along with a pre-populated executionInfo and position instead, we could, but this is the API we agreed on in the KIP.

I want this new API to have good ergonomics, so I do want to consider these, but I don't think we need to hold up the KeyQuery PR on it.

Reply:
I mean, if the upper layers need `QueryResult<R>` with proper type, and the inner layers need `QueryResult<bytes[]>`, we should just have two properly types object, and instead of "swapping", just take the `byte[]` from `QueryResult<bytes[]>`, deserialize them, and stuff the result into the `QueryResult<R>` object.

> but this is the API we agreed on in the KIP.

I did not read the KIP :D (maybe I should have). And we can always adjust it. So me it seems useless to have a generic type parameter if we don't obey it anyway, and use casts. It's the purpose of generics to avoid casts, and if it does not avoid casts, it seems pointless to have).